RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducer.cc

```diff
@@ -8,55 +8,80 @@
         : pfClusParamsToken(esConsumes(config.getParameter<edm::ESInputTag>("pfClusterParams"))),
           topologyToken_(esConsumes(config.getParameter<edm::ESInputTag>("topology"))),
           inputPFRecHitSoA_Token_{consumes(config.getParameter<edm::InputTag>("pfRecHits"))},
           outputPFClusterSoA_Token_{produces()},
           outputPFRHFractionSoA_Token_{produces()},
-          synchronise_(config.getParameter<bool>("synchronise")),
-          pfRecHitFractionAllocation_(config.getParameter<int>("pfRecHitFractionAllocation")) {}
+          num_rhf_{cms::alpakatools::make_host_buffer<uint32_t, Platform>()},
+          synchronise_(config.getParameter<bool>("synchronise")) {}
 
-    void produce(device::Event& event, device::EventSetup const& setup) override {
+    void acquire(device::Event const& event, device::EventSetup const& setup) override {
       const reco::PFClusterParamsDeviceCollection& params = setup.getData(pfClusParamsToken);
       const reco::PFRecHitHCALTopologyDeviceCollection& topology = setup.getData(topologyToken_);
       const reco::PFRecHitHostCollection& pfRecHits = event.get(inputPFRecHitSoA_Token_);
       int nRH = 0;
       if (pfRecHits->metadata().size() != 0)
         nRH = pfRecHits->size();
 
-      reco::PFClusteringVarsDeviceCollection pfClusteringVars{nRH, event.queue()};
-      reco::PFClusteringEdgeVarsDeviceCollection pfClusteringEdgeVars{(nRH * 8), event.queue()};
-      reco::PFClusterDeviceCollection pfClusters{nRH, event.queue()};
-      reco::PFRecHitFractionDeviceCollection pfrhFractions{nRH * pfRecHitFractionAllocation_, event.queue()};
+      pfClusteringVars_ = std::make_optional<reco::PFClusteringVarsDeviceCollection>(nRH, event.queue());
+      pfClusteringEdgeVars_ = std::make_optional<reco::PFClusteringEdgeVarsDeviceCollection>(nRH * 8, event.queue());
+      pfClusters_ = std::make_optional<reco::PFClusterDeviceCollection>(nRH, event.queue());
+
+      *num_rhf_ = 0;
+
+      if (nRH != 0) {
+        PFClusterProducerKernel kernel(event.queue(), pfRecHits);
+        kernel.step1(event.queue(),
+                     params,
+                     topology,
+                     *pfClusteringVars_,
+                     *pfClusteringEdgeVars_,
+                     pfRecHits,
+                     *pfClusters_,
+                     num_rhf_.data());
+      }
+    }
+
+    void produce(device::Event& event, device::EventSetup const& setup) override {
+      const reco::PFClusterParamsDeviceCollection& params = setup.getData(pfClusParamsToken);
+      const reco::PFRecHitHCALTopologyDeviceCollection& topology = setup.getData(topologyToken_);
+      const reco::PFRecHitHostCollection& pfRecHits = event.get(inputPFRecHitSoA_Token_);
+      int nRH = 0;
+      if (pfRecHits->metadata().size() != 0)
+        nRH = pfRecHits->size();
 
       if (nRH != 0) {
+        pfrhFractions_ = std::make_optional<reco::PFRecHitFractionDeviceCollection>(*num_rhf_.data(), event.queue());
```

Comment:
If I read the code correctly, `pfrhFractions_` is used only within the `produce()` method.
In this case it doesn't need to be an `std::optional` data member, it can be a local variable.

Reply:
I think I remember why I made it a member variable. Since the allocation is conditional, I get `error: 'pfrhFractions was not declared in this scope` (note removed the trailing underscore everywhere) on the `event.emplace` further down. Maybe there's an elegant solution?

Reply:
Ah, yes.

It can still be declared as an `std::optional` in the local scope, and assigned like you do now.

Reply:
This was fixed in the latest update.